### ISSUE TITLE
Expose the underlying HTTP error upon `errs.ErrNoMoreHostToTry` failure

### ISSUE THREAD

**Issue by aseure** [algolia/algoliasearch-client-go#579]
- Algolia Client Version: v3.5.2
- Language Version: Go 1.13.7

### Description

The raw error returned after an HTTP request is performed is completely discarded from the end user by [the `transport` layer](https://github.com/algolia/algoliasearch-client-go/blob/v3.5.2/algolia/transport/transport.go#L77-L141). As noted by @marchelbling, it could actually be useful for the end user to access the actual HTTP error.

Here are my first thoughts at the moment.

1. make the error displayed by the `debug` layer; I assume this wouldn't help with production issues which most probably don't run with `debug.Enable()`
2. bubble up the HTTP error with the new Go 1.13 wrapping mechanism; however this is not possible since our support policy is to support the latest N-2 versions of Go at any given time (so we'll be able to use Go 1.13 features once Go 1.15 will be out)
3. create a new function to wrap the `errs.ErrNoMoreHostToTry` error such as [`errs.ErrJSONDecode`](https://github.com/algolia/algoliasearch-client-go/blob/9d7e6def70a5ea50700562d7651e7b572192eed6/algolia/errs/errors.go#L20-L22); however, this would break the compatibility of the public API since `errs.ErrNoMoreHostToTry` is exported

**tl;dr** I'm not yet sure of a backward-compatible way of exposing the HTTP error. I'll try to play with the https://godoc.org/github.com/pkg/errors package (the original implementation of the error wrapping mechanism pre-Go 1.13 which is compatible with our base Go 1.11 version) to see if there's a way to expose the HTTP error while keeping the `errs.ErrNoMoreHostToTry` by default.

### Steps To Reproduce

N/A

**Comment by aseure** [comment]
I've finally found the time to look into this.

In the end, I'm not sure adding [github.com/pkg/errors](https://github.com/pkg/errors) as a dependency is a good idea because:

1. the package is already said to be moved to maintenance mode
2. most importantly, this would force our users to use the `errors.Cause()` function; for Go 1.13 >= users, this means handling errors with both `pkg/errors` and `errors` from the standard library.

As of today, I'd really prefer to make Go 1.13 the minimum version of our library, so that we could use the new error wrapping mechanism from the standard library. I expect most developers to have upgraded to at least Go 1.13 (precisely because of the new error handling mechanism). However, I'm not sure of the impact of the change since I cannot find any meaningful survey about Go version usage among Go developers.

I'd really love to have your thoughts on this @marchelbling @rayrutjes @samouss and other Go developers at Algolia. I'd really love to know more about how painful it is not to be able to access the error. If the pros outweight the cons, let's move on with Go 1.13 and implement a future-proof error handling mechanism.

**Comment by aseure** [comment]
Potentially one more reason to move to Go 1.13: we already use the `go 1.13` directive in our declared `go.mod` file since https://github.com/algolia/algoliasearch-client-go/commit/6be76833adc02e295d2a184d6855a0865485812b made last September.

Not a single user complained about that, so I assume most our users already migrated to Go 1.13 or above.

**Comment by marchelbling** [comment]
Sounds good to me to go for 1.13+!
Having some statistics about go versions usage could be interested but I never came across this kind of survey either.
I can just say that my local `docker` is using go1.12.17 and `kubectl` is go1.12.10 😅 but at least we know that QS is using go1.13+ 😄; I expect that "desktop" binaries will have more inertia than our customers.

**Comment by aseure** [comment]
> I struggle to find what are the errors that are discarded
What we meant here was all the previous errors sent by Algolia. I don't think this is the most important point though, since this error is only supposed to happen when there's a full network disconnection, and we cannot do much about it.

> IMO that could indeed be solved by wrapping the eventual error in go 1.13. Or create, like in some libs, a multierror type that can be examined (like in gorilla schema https://godoc.org/github.com/gorilla/schema#MultiError)

That could be an idea indeed, even though the handling of the error on user-side seems tedious, especially for users whose codebase is already using Go 1.13 error handling mechanism.

After some investigations, I've noticed that the Google Cloud SDK is using build tags and different implementations to handle this. I'll give this solution a try, let me know what you think. For example:

 - [spanner/errors113.go](https://github.com/googleapis/google-cloud-go/blob/master/spanner/errors113.go)
 - [spanner/errors112.go](https://github.com/googleapis/google-cloud-go/blob/master/spanner/errors112.go) (which should be named `errors_not_113.go`)

**Comment by marchelbling** [comment]
@aseure what's the status for this issue? do you need any kind of input to make a decision?

**Comment by aseure** [comment]
@marchelbling I'm gonna work on that again this week. Here's the plan:

 - Try to implement something for Go <1.13 and Go >=1.13 using build tags
 - If this leads to important drawbacks for the end-user, we're finally gonna upgrade to Go 1.13 and use the standard library error wrapping mechanism.
